gix-config/src/parse/nom/mod.rs

```diff
@@ -158,303 +156,219 @@ fn section_header(i: &[u8]) -> IResult<&[u8], section::Header<'_>> {
+                dispatch(Event::Value(Cow::Borrowed("".into())));
+                return Ok(());
             } else {
-                dispatch(Event::Value(Cow::Borrowed(value.as_bstr())));
+                last_value_index
             }
-        };
-        (&i[value_start..], value_end - value_start, newlines, dispatch)
-    };
-
-    let (i, remainder_value) = {
-        let value_end_no_trailing_whitespace = i[..value_end]
-            .iter()
-            .enumerate()
-            .rev()
-            .find_map(|(idx, b)| (!b.is_ascii_whitespace()).then_some(idx + 1))
-            .unwrap_or(0);
-        (
-            &i[value_end_no_trailing_whitespace..],
-            &i[..value_end_no_trailing_whitespace],
-        )
+        }
+        Some(idx) => idx,
     };
 
-    dispatch(remainder_value);
+    i.reset(value_start_checkpoint);
+    let value_end_no_trailing_whitespace = i[..value_end]
+        .iter()
+        .enumerate()
+        .rev()
+        .find_map(|(idx, b)| (!b.is_ascii_whitespace()).then_some(idx + 1))
+        .unwrap_or(0);
+    let remainder_value = i.next_slice(value_end_no_trailing_whitespace);
+
+    if partial_value_found {
+        dispatch(Event::ValueDone(Cow::Borrowed(remainder_value.as_bstr())));
+    } else {
+        dispatch(Event::Value(Cow::Borrowed(remainder_value.as_bstr())));
+    }
 
-    Ok((i, newlines))
+    Ok(())
 }
 
-fn take_spaces(i: &[u8]) -> IResult<&[u8], &BStr> {
-    let (i, v) = take_while(|c: u8| c.is_ascii() && is_space(c))(i)?;
-    if v.is_empty() {
-        Err(nom::Err::Error(NomError {
-            input: i,
-            code: ErrorKind::Eof,
-        }))
-    } else {
-        Ok((i, v.as_bstr()))
-    }
+fn take_spaces1<'i>(i: &mut &'i [u8]) -> PResult<&'i BStr, NomError<&'i [u8]>> {
+    take_while(1.., winnow::stream::AsChar::is_space)
+        .map(bstr::ByteSlice::as_bstr)
+        .parse_next(i)
 }
 
-fn take_newlines(i: &[u8]) -> IResult<&[u8], (&BStr, usize)> {
-    let mut counter = 0;
-    let mut consumed_bytes = 0;
-    let mut next_must_be_newline = false;
-    for b in i.iter().copied() {
-        if !b.is_ascii() {
-            break;
-        };
-        if b == b'\r' {
-            if next_must_be_newline {
-                break;
-            }
-            next_must_be_newline = true;
-            continue;
-        };
-        if b == b'\n' {
-            counter += 1;
-            consumed_bytes += if next_must_be_newline { 2 } else { 1 };
-            next_must_be_newline = false;
-        } else {
-            break;
-        }
-    }
-    let (v, i) = i.split_at(consumed_bytes);
-    if v.is_empty() {
-        Err(nom::Err::Error(NomError {
-            input: i,
-            code: ErrorKind::Eof,
-        }))
-    } else {
-        Ok((i, (v.as_bstr(), counter)))
-    }
+fn take_newlines1<'i>(i: &mut &'i [u8]) -> PResult<&'i BStr, NomError<&'i [u8]>> {
+    repeat(1.., alt(("\r\n", "\n")))
+        .map(|()| ())
```

Comment:
btw I figure this should be called out.

Nom's `many` always allocated into a `Vec`.

Winnow's `repeat` instead allocates into `trait Accumulate` which is kind of like `Default + Extend` with a `with_capacity` **but** we implement it for `usize` (to count parses) and `()` to ignore the output o the parser, removing the need for `count_` and `skip_` variants of these parsers.

`map(|()| ())` is a quick way of annotating the `repeat` which I feel is nicer than `repeat<_, _, (), _>` (or whatever the generic params are).  I'd love to have a `Parser::as_void` to make this more explicit but haven't figured out a way to do that.

Reply:
Thanks for pointing it out! It's definitely a major improvement already I can see why `winnow` is intriguing to (even) switch to.
And it's even *just* the beginning :).